File: solutions/cpp/16113/main.cpp
#include<bits/stdc++.h>

using namespace std;

vector<vector<string>> digits = {
    {"###","#.#","#.#","#.#","###",},
    {"#","#","#","#","#",},
    {"###","..#","###","#..","###",},
    {"###","..#","###","..#","###",},
    {"#.#","#.#","###","..#","..#",},
    {"###","#..","###","..#","###",},
    {"###","#..","###","#.#","###",},
    {"###","..#","..#","..#","..#",},
    {"###","#.#","###","#.#","###",},
    {"###","#.#","###","..#","###",},
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int N; cin >> N;
    string S; cin >> S;
    int width = N / 5;
    vector<string> signal;
    for(int i = 0; i < N; i += width) {
        signal.emplace_back(S.substr(i, width));
    }
    vector<int> idx;
    idx.push_back(-1);
    for(int i = 0; i < width; ++i) {
        bool flag = true;
        for(int j = 0; j < 5; ++j) if(signal[j][i] != '.') flag = false;
        if(flag) idx.push_back(i);
    }
    idx.push_back(width);
    vector<int> answer;
    for(int i = 1; i < (int)idx.size(); ++i) {
        int l = idx[i - 1], r = idx[i]; // [l + 1, r - 1]
        if(r - l - 1 <= 0) continue;
        for(int digit = 0; digit < 10; ++digit) {
            bool flag = true;
            for(int j = 0; j < 5; ++j) {
                if(digits[digit][j] != signal[j].substr(l + 1, r - l - 1)) {
                    flag = false;
                }
            }
            if(flag) {
                answer.push_back(digit);
                break;
            }
        }
    }
    for(int x : answer) cout << x;

    return 0;
}

File: solutions/python/16113/main.py
import sys
from typing import List


def input() -> str:
    return sys.stdin.readline().rstrip()


Digits: List[List[str]] = [
    ["###", "#.#", "#.#", "#.#", "###"],
    ["#", "#", "#", "#", "#"],
    ["###", "..#", "###", "#..", "###"],
    ["###", "..#", "###", "..#", "###"],
    ["#.#", "#.#", "###", "..#", "..#"],
    ["###", "#..", "###", "..#", "###"],
    ["###", "#..", "###", "#.#", "###"],
    ["###", "..#", "..#", "..#", "..#"],
    ["###", "#.#", "###", "#.#", "###"],
    ["###", "#.#", "###", "..#", "###"],
]


def solution(N: int, S: str) -> List[int]:
    w = len(S) // 5
    signal: List[str] = [S[i : i + w] for i in range(0, N, w)]
    idx: List[int] = [-1]
    for i in range(w):
        flag = True
        for j in range(5):
            if signal[j][i] != ".":
                flag = False
        if flag:
            idx.append(i)
    idx.append(w)
    answer: List[int] = []
    for i in range(1, len(idx)):
        l, r = idx[i - 1], idx[i]
        if r - l - 1 <= 0:
            continue
        for digit in range(10):
            flag = True
            for j in range(5):
                if Digits[digit][j] != signal[j][l + 1 : r]:
                    flag = False
            if flag:
                answer.append(digit)
                break
    return answer


if __name__ == "__main__":
    N = int(input())
    S = input()
    answer = solution(N, S)
    answer = "".join(map(str, answer))
    print(answer)
